### ISSUE TITLE
The camera transformation matrix

### ISSUE THREAD

**Issue by Jianf-Wang** [mjvanderboon/headnerf#1]
Hi, I am also using DECA. Have you found the way to estimate the extrinsic matrix (3 * 4) with DECA ?  The code_dict['pose'] only returns  six parameters. Thanks

**Comment by mjvanderboon** [comment]
Hi Jianf-Wang, thanks for your question.

I used some available functionality in DECA to convert their code_dict['pose'] to a rotation matrix. I then use code_dict['cam'] to generate a homogeneous transformation matrix from world 2 camera frame. I matrix multiply these two transformations and use them for the final homogeneous transformation of the face (to camera frame). The relevant code is [here](https://github.com/mjvanderboon/headnerf/blob/66c0d1e6868fce7dfa2f48c861e2f875a8f4d33a/im2scene/headnerf/training.py#L98). IIRC the code_dict['pose'][0:3] holds an axis angle representation of the pose of the head, the rest code_dict['pose'][3:] are joint angles. 

I never got around to checking if this code works as I expect. If you experiment more with DECA, please let me know if this approach results in the correct homogeneous transformation matrix. I also opened an issue regarding this problem in the DECA repo [here](https://github.com/YadiraF/DECA/issues/110).

Also, if you are interested in contributing to this repository to develop the code publicly that is described in the headnerf paper, let me know! We can schedule a call.

**Comment by Jianf-Wang** [comment]
Hi, Thanks for your response. I am using DECA for another task, and I am confused about how you use code_dict['cam']. Is it a set of translation parameters ?  what does scale mean ? (I think it should be translation (x_t, y_t, z_t), but it returns (x_t, y_t, scale)).

**Comment by tomguluson92** [comment]
> Hi, Thanks for your response. I am using DECA for another task, and I am confused about how you use code_dict['cam']. Is it a set of translation parameters ? what does scale mean ? (I think it should be translation (x_t, y_t, z_t), but it returns (x_t, y_t, scale)).

camera intrinsic, `x_t`, `y_t` are the offset on x, y of camera plane, `scale` is the scale coefficient, with regard to the focal (with no distortion).

@ref: https://towardsdatascience.com/camera-intrinsic-matrix-with-example-in-python-d79bf2478c12

**Comment by tomguluson92** [comment]
Hi @mjvanderboon , I have test your transformation code, which is an expected result!
![Dingtalk_20220426142320](https://user-images.githubusercontent.com/19585240/165235264-1b208d5b-ade5-4dbe-9f40-3e84a987f25d.jpg)

camera1 -->
![007999_vis_original_size](https://user-images.githubusercontent.com/19585240/165235375-24e190f2-a0e2-4dec-9fdc-0130525972a7.jpg)


camera2 -->
![007000_vis_original_size](https://user-images.githubusercontent.com/19585240/165235386-b6e428c7-ec10-4378-8c91-23859e4a43e0.jpg)

**Comment by mjvanderboon** [comment]
Hi @tomguluson92, thanks for your reply to this issue. Good to hear the logic works as expected! For reference for future use, could you run me through how you generated the images you attached and what they show?

**Comment by tomguluson92** [comment]
> Hi @tomguluson92, thanks for your reply to this issue. Good to hear the logic works as expected! For reference for future use, could you run me through how you generated the images you attached and what they show?

Do you use Discord? You could leave me your ID

**Comment by mjvanderboon** [comment]
If you're able, I'd prefer a short comment in this issue so other people can find it in the future. Also for others who use DECA!

**Comment by tomguluson92** [comment]
``` python
import scipy.io
import torch
from camera_visualization import plot_camera_scene
from pytorch3d.renderer.cameras import (
    SfMPerspectiveCameras,
)

device = torch.device("cuda:0")

content1 = scipy.io.loadmat("./data/007999/007999.mat")
content2 = scipy.io.loadmat("./data/007000/007000.mat")

pose1 = torch.from_numpy(content1['pose'])
pose2 = torch.from_numpy(content2['pose'])

rotation_matrices1 = batch_rodrigues(pose1[:, :3])
rotation_matrices2 = batch_rodrigues(pose2[:, :3])

R_deca1 = rotation_matrices1
R_deca2 = rotation_matrices2

T_deca1 = R_deca1[:, :, -1]
T_deca2 = R_deca2[:, :, -1]

cameras_deca1 = SfMPerspectiveCameras(
    R = R_deca1.to(device),
    T = T_deca1.to(device),
    device = device,
)
cameras_deca2 = SfMPerspectiveCameras(
    R = R_deca2.to(device),
    T = T_deca2.to(device),
    device = device,
)

plot_camera_scene(cameras_deca1, cameras_deca2, "20220426_deca")
```

`plot_camera_scene` comes from the file:
``` bash
!wget https://raw.githubusercontent.com/facebookresearch/pytorch3d/main/docs/tutorials/utils/camera_visualization.py
from camera_visualization import plot_camera_scene
```

**Comment by mjvanderboon** [comment]
Great! Thanks for adding the script. I'm closing this issue now.

**Comment by Jianf-Wang** [comment]
Hi, thanks for your response and help !

**Comment by sunshineatnoon** [comment]
Hi @tomguluson92 Thanks for the script. I am confused about `T_deca1 = R_deca1[:, :, -1]`, as far as I understand, the `batch_rodrigues` function only return the rotation matrix, so how can we get the 3D translation from `R_deca1`?

**Comment by sunshineatnoon** [comment]
Also, is there any way to get the intrinsics parameters from DECA?

**Comment by emlcpfx** [comment]
Has anyone written code that exports the DECA camera to Blender, C4D, or similar 3D software?
